File: flyweight/python/MazeGame.py
import random

class MazeGame:
    def __init__(self):
        self.componentFactory = ComponentFactory()

    def createMaze(self, x, y):
        m = self.makeMaze(x, y)

        for i in range(x):
            for j in range(y):
                m.addComponent(self.makeWall(), i, j)

        m.addComponent(self.makeRoom(), 1, 2)
        m.addComponent(self.makeRoom(), 2, 2)
        m.addComponent(self.makeRoom(), 3, 2)
        m.addComponent(self.makeRoom(), 1, 1)
        m.addComponent(self.makeRoom(), 3, 1)
        m.addComponent(self.makeRoom(), 1, 3)
        m.addComponent(self.makeRoom(), 3, 3)

        return m

    def makeMaze(self, x, y):
        return Maze(x, y)

    def makeRoom(self):
        return self.componentFactory.createComponent("coinroom")

    def makeWall(self):
        return self.componentFactory.createComponent("wall")

class Maze:
    def __init__(self, x, y):
        self.seed = random.randint(0, 5000)
        self.length = x
        self.height = y
        self.grid = []

        for i in range(x):
            self.grid.append([])
            for j in range(y):
               self.grid[i].append(0) 

    def addComponent(self, component, x, y):
        self.grid[x][y] = component

    def printMaze(self):
        context = MazeContext(self.seed)
        for y in range(self.length):
            for x in range(self.height):
                print(self.grid[x][y].getSymbol(context), end="")
                context.increment()
            print("")


class ComponentFactory:
    def __init__(self):
        self.flyweights = {} 

    def createComponent(self, componentType):
        if componentType not in self.flyweights:
            self.flyweights[componentType] = MazeComponent(componentType)

        return self.flyweights[componentType]

class MazeComponent:
    def __init__(self, ct):
        self.componentType = ct

    def isSolid(self):
        if self.componentType == "wall":
            return True
        else:
            return False

    def getSymbol(self, mazeContext):
        if self.componentType == "wall":
            return '#'
        elif self.componentType == "room":
            return '.'
        elif self.componentType == "coinroom":
            return '@' if mazeContext.hasCoin() else '.'
        else:
            return ' '

class MazeContext:
    def __init__(self, seed):
        self.seed = seed
        self.index = 0

    def increment(self):
        self.index += 1

    def hasCoin(self):
        random.seed(self.seed*self.index)
        coin = False
        if random.randint(0, 100) >= 75:
            coin = True
        
        return coin

def main():
   mg = MazeGame() 
   m = mg.createMaze(5, 5)
   m.printMaze()

if __name__ == "__main__":
    main()

File: strategy/js/MazeGame.js
function MazeGame(){
}

MazeGame.createMaze = function(mf) {
	if (!(mf instanceof MazeFactory)) {
		throw new Error("Invalid MazeFactory used!");
	}
	var m = mf.makeMaze(5, 5);

	for (var x = 0; x < 5; x++) {
		for (var y = 0; y < 5; y++) {
			m.addComponent(mf.makeWall(), x, y);
		}
	}
	m.addComponent(mf.makeRoom(), 1, 2);
	m.addComponent(mf.makeRoom(), 2, 2);
	m.addComponent(mf.makeRoom(), 3, 2);
	m.addComponent(mf.makeRoom(), 1, 1);
	m.addComponent(mf.makeRoom(), 3, 1);
	m.addComponent(mf.makeRoom(), 1, 3);
	m.addComponent(mf.makeRoom(), 3, 3);

	return m;
}

function MazeFactory() {
	this.makeMaze = function(x, y) {
		return new Maze(x, y);
	}

	this.makeRoom = function() {
		return new Room();
	}

	this.makeWall = function() {
		return new Wall();
	}
}

function CoinFactory() {
	MazeFactory.apply(this, arguments);

	this.makeRoom = function() {
		return new CoinRoom();
	}
}
CoinFactory.prototype = new MazeFactory();

function Maze(x, y) {
	var length = x;
	var height = y;
	var map = new Array(x);

	for (var i = 0; i < x; i++) {
		map[i] = new Array(y);
		for (var j = 0; j < y; j++) {
			map[i][j] = 0;
		}
	}

	this.addComponent = function(component, x, y) {
		if (!(x >= length || y >= height || x < 0 || y < 0)) {
			map[x][y] = component;
		}
	}

	this.printMaze = function() {
		for (var y = 0; y < 5; y++) {
			for (var x = 0; x < 5; x++) {
				process.stdout.write(map[x][y].getSymbol());
			}
			process.stdout.write("\n");
		}
	}
}

function Room() {
	this.isSolid = function() {
		return false;
	}

	this.getSymbol = function() {
		return '.';
	}
}

function CoinRoom() {
	Room.apply(this, arguments);
	var coin = false;
	if (Math.random() >= 0.75) {
		coin = true;
	}

	this.getSymbol = function() {
		if (coin == true) {
			return '@';
		} else {
			return '.';
		}
	}
}
CoinRoom.prototype = new Room();

function Wall() {
	this.isSolid = function() {
		return true;
	}

	this.getSymbol = function() {
		return '#';
	}
}

var mf = new CoinFactory();
var m = MazeGame.createMaze(mf);
m.printMaze();
